### PR TITLE
Lint for Result/Option.unwrap()

### LINKED ISSUE #24: `unwrap()` of Option<_> or Result<_>
I believe we should warn about usage of `unwrap()` of optional types. `unwrap()` is essentially a pattern match with `_ => panic!()`, but there are often ways to continue execution of program w/o failing.

Today, it's common to find code like

```Rust
extern crate term;

fn main() {
    let mut t = term::stdout().unwrap();

    t.fg(term::color::GREEN).unwrap();
    (write!(t, "hello, ")).unwrap();

    t.fg(term::color::RED).unwrap();
    (writeln!(t, "world!")).unwrap();

    t.reset().unwrap();
}
```

As one might learn pretty quickly, `unwrap()`ing a None returned by `term::stdout()` will lead to failure of task. This is, of course, specified in docs. But it's certainly not obvious that this is not a "safe" operation — it may lead to "crash". Not "hard" crash, because it will be handled by Rust runtime. It produces error about unwrapping None. But it's a crash nevertheless.

Warnings from compiler that `.unwrap()` is essentially a pattern match w/o "catch-all" case are missing here. We do get warnings when not all variants of enum are handled, so why skip it here?

`unwrap()` of `Option<_>` is a hack. It's convenient, I know — but it's the sort of convenience we disdain JavaScript and PHP for. I also understand that one is extremely likely to use `unwrap()` in closures because "heck, I only need to filter/map/reduce stuff, I don't want my closure to spend 5 lines!". Programmer uses it there and then forgets about it. Voila — safety is needlessly violated.

I think a programmer should be forced to actively discard any failure condition. It makes for explicit decision making about not handling an error.

In Rust, in it's present state, `unwrap()` is the easiest and simplest way to Get Things Done. This shortcut will be taken by many and many people who will then later complain about Rust not being secure despite claiming itself "preventing almost all crashes".

### PR DESCRIPTION
**Pull request by birkenfeld**
Fixes #24

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -84,3 +84,11 @@ pub fn span_help_and_lint(cx: &Context, lint: &'static Lint, span: Span,
         cx.sess().span_help(span, help);
     }
 }
+
+/// return the base type for references and raw pointers
+pub fn walk_ty<'t>(ty: ty::Ty<'t>) -> ty::Ty<'t> {
```
Comment (Manishearth) [inline src/utils.rs]: Can we rename this to `walk_ptrs_ty` or something instead? There's another `walk_ty` in `middle` which is quite useful and could get confusing.

File: src/lib.rs [hunk 2]
```diff
@@ -29,6 +29,7 @@ pub mod collapsible_if;
 pub mod unicode;
 pub mod utils;
 pub mod strings;
+pub mod unwrap;
```
Comment (Manishearth) [inline src/lib.rs]: Could we call this methods instead?

File: src/unwrap.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
+use syntax::ast::*;
+use rustc::lint::{Context, LintPass, LintArray};
+use rustc::middle::ty;
+
+use utils::{span_lint, match_def_path, walk_ty};
+
+#[derive(Copy,Clone)]
+pub struct UnwrapPass;
```
Comment (Manishearth) [inline src/unwrap.rs]: MethodsPass

File: src/unwrap.rs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+              "Warn on using unwrap() on a Result value");
+
+impl LintPass for UnwrapPass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(OPTION_UNWRAP_USED, RESULT_UNWRAP_USED)
+    }
+
+    fn check_expr(&mut self, cx: &Context, expr: &Expr) {
+        if let ExprMethodCall(ref ident, _, ref args) = expr.node {
+            if ident.node.name != "unwrap" {
```
Comment (Manishearth) [inline src/unwrap.rs]: nit: move the rest of the fn to inside this `if` so that it's visibly easy to extend for other method checks

**Comment by Manishearth** [comment]
LGTM, some minor changes needed

**Comment by Manishearth** [comment]
(Oh, also, it's worthwhile to have `(fixes #24)` in one of your commit messages so GitHub auto-closes the issue)

**Comment by birkenfeld** [comment]
I think I addressed all the comments.

**Comment by Manishearth** [comment]
Thanks!

**Pull request merged**